flow/StreamCipher.cpp

```diff
@@ -19,75 +19,83 @@
  */
 
 #include "flow/StreamCipher.h"
+#include "flow/Arena.h"
+#include "flow/ITrace.h"
 #include "flow/UnitTest.h"
+#include <memory>
 
 std::unordered_set<EVP_CIPHER_CTX*> StreamCipher::ctxs;
-std::unique_ptr<StreamCipher::Key> StreamCipher::Key::globalKey;
+std::unique_ptr<StreamCipherKey> StreamCipherKey::globalKey;
 
-StreamCipher::StreamCipher() : ctx(EVP_CIPHER_CTX_new()) {
-	ctxs.insert(ctx);
+StreamCipherKey* StreamCipherKey::getGlobalCipherKey() {
+	if (StreamCipherKey::globalKey.get() == nullptr) {
+		globalKey = std::make_unique<StreamCipherKey>(AES_256_KEY_LENGTH);
+		return globalKey.get();
+	}
+	return globalKey.get();
 }
 
-StreamCipher::~StreamCipher() {
-	EVP_CIPHER_CTX_free(ctx);
-	ctxs.erase(ctx);
+void StreamCipherKey::setGlobalCipherKey(uint8_t* const data, int len) {
+	StreamCipherKey* cipherKey = StreamCipherKey::getGlobalCipherKey();
+	cipherKey->initializeKey(data, len);
 }
 
-EVP_CIPHER_CTX* StreamCipher::getCtx() {
-	return ctx;
-}
-
-void StreamCipher::cleanup() noexcept {
-	Key::cleanup();
-	for (auto ctx : ctxs) {
-		EVP_CIPHER_CTX_free(ctx);
+void StreamCipherKey::globalKeyCleanup() {
+	if (StreamCipherKey::globalKey != nullptr) {
+		StreamCipherKey* cipherKey = StreamCipherKey::getGlobalCipherKey();
+		cipherKey->cleanup();
 	}
 }
 
-void StreamCipher::Key::initializeKey(RawKeyType&& arr) {
-	if (globalKey) {
-		ASSERT(globalKey->arr == arr);
-	}
-	globalKey = std::make_unique<Key>(ConstructorTag{});
-	globalKey->arr = std::move(arr);
-	memset(arr.data(), 0, arr.size());
+StreamCipher::StreamCipher(int keySize)
```

Comment:
Isn't the key size always `AES_256_KEY_LENGTH`? The `keySize` parameter could be removed here to improve encapsulation.

Reply:
For now it is AES 256, however, upgrading protocol further, the key-size can be changed as well. Hence, the API takes keySize as the parameter.